### PR TITLE
Supported the 4D shapes

### PR DESCRIPTION
**Pull request by Wenzhao-Xiang**
#95 4D shapes are supported now.

### CODE REVIEW

**Comment by huningxin** [comment]
Thanks for the PR @Wenzhao-Xiang ! I would suggest to change the commit log, e.g. 
```
[WebGL2] support 4D shapes

Fix #95
```

Please refer to https://help.github.com/articles/closing-issues-using-keywords/ for more details.

File: src/nn/Execution.js [hunk 1]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
    * Schedule evaluation of the execution.
    */
   async startCompute() {
-    await this._preparedModel.execute(this._inputs, this._outputs);
+    let input_shape = this._model._operands[this._inputs.get(0).index].dimensions;
```
Comment (huningxin) [inline src/nn/Execution.js]: Use `const` if possible

File: src/nn/Execution.js [hunk 2]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
...
+    if (input_shape.length === 4){
+      let output_shape = this._model._operands[this._outputs.get(0).index].dimensions;
+      let buffer = this._inputs.get(0).buffer;
+      let input_size = input_shape[1] * input_shape[2] * input_shape[3];
```
Comment (huningxin) [inline src/nn/Execution.js]: do we need to assert the batch size must be 1?

File: src/nn/Execution.js [hunk 3]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
...
+      let output_size = output_shape[1] * output_shape[2] * output_shape[3];
```
Comment (huningxin) [inline src/nn/Execution.js]: what about the output shape is not 4D?

File: src/nn/Execution.js [hunk 4]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
...
+      let inputIndex = this._inputs.get(0).index;
+      let outputIndex = this._outputs.get(0).index;
```
Comment (huningxin) [inline src/nn/Execution.js]: how do we handle multiple inputs and outputs case?

File: src/nn/Execution.js [hunk 5]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
...
+      let outputBuffer = [];
+      let tmp_inputs = new Map();
+      let tmp_outputs = new Map();
+      for(let i=0;i < input_shape[0]; ++i){
```
Comment (huningxin) [inline src/nn/Execution.js]: add space between ops, e.g. `for (let i = 0; i < input_shape[0]; ++i)`

File: src/nn/Execution.js [hunk 6]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
...
+        tmp_inputs.set(0,{index: inputIndex, buffer: buffer.slice(input_size * i,input_size * (i+1))});
```
Comment (huningxin) [inline src/nn/Execution.js]: add space after `,`

File: src/nn/Execution.js [hunk 7]
```diff
@@ -82,7 +82,28 @@ export default class Execution {
...
+        tmp_outputs.set(0,{index: outputIndex, buffer: buffer.slice(output_size * i,output_size * (i+1))});
+        await this._preparedModel.execute(tmp_inputs, tmp_outputs);
```
Comment (huningxin) [inline src/nn/Execution.js]: do we need this code for wasm path? #95 is for WebGL2, did we test wasm?

Comment (Wenzhao-Xiang) [reply src/nn/Execution.js]: no,wasm support the 4D shapes.
I have optimized the code.

**Comment by huningxin** [comment]
@GreyZzzzzzXh , please help take a look at this PR. Thanks!

**Comment by GreyZzzzzzXh** [comment]
@huningxin  We reviewed the code and did some changes including coding style. It works well in all test cases and examples now.

**Comment by huningxin** [comment]
Looks good. Thanks for the PR and review. Merging.

**Pull request merged**